### ISSUE TITLE
Add Length property for arrays

### ISSUE THREAD

**Issue by alkhimey** [adventuregamestudio/ags#637]
Currently, the workaround for not being able to "ask" an array of it's length is to keep track of the length using separate variable. This is inconvenient and can lead to nasty bugs when the length is changed, but the variable is not.

Instead of:

    int a[] = new int[3];
    readonly int len = 3; // should be same as the length of a
    f(a, len);

    function(int[] a; int len) {
        for (i=0; i < len; i++) {
            ...
        }
    }

We could:

    int a[] = new int[3];
    f(a);

    function(int[] a) {
        for (i=0; i < a.Length; i++) {
            ...
        }
    }

**Comment by fernewelten** [comment]
As I see it, there are two main ways to do it.

__One way__ would be to get the Engine to help. It already knows the size of dynamic memory blocks and uses this for memory overrun checks (opcode `SCMD_DYNAMICBOUNDS`). So we could add an opcode that reads the memory size and returns it into a register. The compiler would then translate all calls to `array.Length` into code that uses the new opcode.

So the disadvantage is that the Engine would need to be extended.

__The other way__ would do all the logic on the compiler side, with the existing opcodes: 
- Whenever a program defines a dynamic array `a`, the compiler would secretly add a definition for `a#Length`. 
- Whenever the program does `a = new Struct[size];`, the compiler would generate the equivalent for `a#Length = size;`.
- Whenever a program defines a function with a dynamic array parameter, the compiler would add a secret length parameter. And when the program passes `a` to the function, the compiler would pass `a#Length` to the secret extra parameter. 
- Whenever a program defines a function that returns a dynamic array, then the corresponding bytecode would return the array in register AX and its size in BX. The caller of such a function would read BX after it returns and use this value to keep track of the size of the returned array.

The disadvantages: 
- All existing system functions that either accept dynamic array parameters or return dynamic arrays would need to be rewritten to accommodate the new calling convention.  
- __Both__ the old and the new compiler would need to be modified in parallel.  Otherwise, since the calling convention for functions is changed, the old compiler couldn't compile _any_ program any longer, no matter whether that program uses `a.Length`.

**Comment by ivan-mogilko** [comment]
Ermmm... yeah, I would go with engine's approach :).
I mean, dynamic arrays are already *managed objects*, so that's natural that engine participates in handling them.

**Comment by ivan-mogilko** [comment]
Hey, but, in fact, there's another 3rd option:

* don't add any new opcodes;
* define a hard-coded *function name* accepting 1 argument and returning int. Don't actually declare it, as agsscript does not support generic values, just pretend that it exists;
* when array.Length syntax is met, add call to that function, passing array mem pointer as an argument;
* implement this function in the engine; treat the single argument as a pointer to dynamic array object, and return its length.

**Comment by fernewelten** [comment]
I like that third option. I haven't looked into how system function names are mapped to C++ function names. Perhaps we can get away with a function name that just __can't__ be defined within AGS code, e.g., by embedding a non-alphanumeric character in it? Something like `array$length` or `_#ArrayLength`.

**Comment by rofl0r** [comment]
my favorite would be `__builtin_length(object)` along the lines of how GCC does such things. it's very unlikely someone used this exact name in his code in the past.

**Comment by ericoporto** [comment]
I may be remembering wrong, but I thought we now had multiple dimensions in arrays, if that's a thing Length needs an argument for the dimension (default 0), and it needs a rank to figure out how many dimensions the array has.

**Comment by fernewelten** [comment]
I've only implemented non-dynamic multidimensional arrays for now because doing this for dynamic arrays opens another can of worms. The Engine currently doesn't track multiple dimensions and it currently doesn't grok pointers of pointers.

But yes, if we have a `Length` operator for dynamic arrays, it should work also for multidimensional static arrays, dimension by dimension. In this case, this is easy for the compiler to answer because the dimensions concerned are fixed and known at compile time. They are part of the type system in use and tracked in the symbol table. After the statement `int var[2][3];`, `var` has the type `int[2][3]`, and the symbol table for `int[2][3]` has a vector containing ´{2, 3}`. So in that case, no need to call any external function.

**Comment by ericoporto** [comment]
Just somewhat of what I thought from other script languages:
```
int var[4][3];
var.Length; //returns 4
var[0].Length; //returns 3
var.Size; //returns [4, 3]
var.Size[0]; //returns 4
var.Size[1]; //returns 3
var.Rank; //returns 2
```

**Comment by ivan-mogilko** [comment]
I'm afraid we might not have enough type support for something like above.

If you take `var[0].Length` for example, this means `var[0]` should be a meaningful object, but is it now?

Also, how (in which form) do you return something like [4, 3]?

**Comment by ivan-mogilko** [comment]
Perhaps make simply .Length[x] for static arrays, as fernewelten proposed above, for now, as static arrays are fixed in sizes so there's no need to test individual rows & columns for their size.

And .Length for dynamic array. If there will be support for multidimensional dynamic arrays, this will still work to test the size of each individual array on every rank.

**Comment by fernewelten** [comment]
I'm addressing this for dynamic arrays in #1256.

We don't really need this for static arrays right now because those can't be passed to functions directly. You can't write a function that handles "all kinds of static arrays, no matter what their size". So there's no urgent need to query the size of static arrays at runtime.

**Comment by ivan-mogilko** [comment]
Resolved by #1256 in the new script compiler.
